project/apps/web/app/search/page.tsx

```diff
@@ -0,0 +1,237 @@
+    onMutate: () => {
+      startSearch();
+    },
+    onSuccess: (data) => {
+      setMatchReqId(data.match_req_id);
+    },
+    onError: (error: any) => {
+      stopSearch();
+      toast({
+        variant: 'error',
+        title: 'Error',
+        description: error.message,
+      });
+    },
+  });
+
+  const cancelMatchMutation = useMutation({
+    mutationFn: (matchCancel: MatchCancelDto) => cancelMatch(matchCancel),
+    onSuccess: () => {
+      toast({
+        title: 'Canceled',
+        description: 'Your match request has been successfully canceled.',
+      });
+    },
+    onError: (error: any) => {
+      toast({
+        variant: 'error',
+        title: 'Error',
+        description: `Failed to cancel match: ${error.message}`,
+      });
+    },
+  });
+
+  const handleCreateMatch = (newMatch: MatchRequestMsgDto) => {
+    setTimer(0);
+    createMutation.mutate(newMatch);
+  };
+
+  const stopMatching = () => {
+    if (intervalRef.current !== null) {
+      clearInterval(intervalRef.current);
+      intervalRef.current = null;
+    }
+    if (matchReqId) {
+      const cancelMatchReq: MatchCancelDto = { match_req_id: matchReqId };
+      cancelMatchMutation.mutate(cancelMatchReq);
+    }
+    stopSearch();
+  };
+
+  // Set up WebSocket event listeners for match notifications
+  useEffect(() => {
+    if (!socket) return;
+
+    const handleMatchFound = (message: string) => {
+      console.log('Match found:', message);
+      setMatchFound(true);
+      stopSearch();
+      toast({
+        variant: 'success',
+        title: 'Match Found',
+        description: 'Your match was successful.',
+      });
+
+      router.push(`/match/${message}`);
+    };
+
+    const handleMatchInvalid = (message: string) => {
+      console.log('Match invalid:', message);
+      stopSearch();
+      stopMatching();
+      toast({
+        variant: 'error',
+        title: 'Match Invalid',
+        description: 'Your match was invalid.',
+      });
+    };
+
+    const handleMatchRequestExpired = (message: string) => {
+      console.log('Match request expired:', message);
+      stopSearch();
+      stopMatching();
+      toast({
+        variant: 'error',
+        title: 'Match Expired',
+        description: 'Your match request has expired. Please try again.',
+      });
+    };
+
+    socket.on('match_found', handleMatchFound);
+    socket.on('match_invalid', handleMatchInvalid);
+    socket.on('match_request_expired', handleMatchRequestExpired);
+
+    return () => {
+      socket.off('match_found', handleMatchFound);
+      socket.off('match_invalid', handleMatchInvalid);
+      socket.off('match_request_expired', handleMatchRequestExpired);
+    };
+  }, [socket, stopSearch, toast, router]);
+
+  // Handle match creation from query params
+  useEffect(() => {
+    if (matchDataParam) {
+      const matchData: MatchRequestMsgDto = JSON.parse(matchDataParam);
+      handleCreateMatch(matchData);
+    }
+  }, [matchDataParam]);
```

Comment:
It's only being called twice because of [react strict mode](https://nextjs.org/docs/pages/api-reference/next-config-js/reactStrictMode), which is enabled by default using our version of Next.

However, you are right that if there is the case whereby this useEffect runs twice because of side effects, we should handle it as well - I will add a check for it.